Assets/LeapMotionModules/InteractionEngine/Scripts/ControllerImplementations/HeuristicGrabClassifier.cs

```diff
@@ -0,0 +1,80 @@
+﻿using UnityEngine;
+using System.Collections.Generic;
+using InteractionEngineUtility;
+
+namespace Leap.Unity.Interaction {
+  public class HeuristicGrabClassifier {
+    Collider[] collidingCandidates = new Collider[10];
+    public InteractionManager _manager;
+
+    Dictionary<IInteractionBehaviour, GrabClassifierHeuristics.GrabClassifier> leftGrabClassifiers = new Dictionary<IInteractionBehaviour, GrabClassifierHeuristics.GrabClassifier>();
+    Dictionary<IInteractionBehaviour, GrabClassifierHeuristics.GrabClassifier> rightGrabClassifiers = new Dictionary<IInteractionBehaviour, GrabClassifierHeuristics.GrabClassifier>();
+    GrabClassifierHeuristics.ClassifierParameters grabParams;
+
+
+    public HeuristicGrabClassifier(InteractionManager manager, float fingerStickiness = 0f, float thumbStickiness = 0.05f, float maxCurl = 0.65f, float minCurl = -0.1f, float fingerRadius = 0.01f, float thumbRadius = 0.015f, float grabCooldown = 0.2f) {
+      _manager = manager;
+      grabParams = new GrabClassifierHeuristics.ClassifierParameters(fingerStickiness, thumbStickiness, maxCurl, minCurl, fingerRadius, thumbRadius, grabCooldown);
+    }
+
+    public void UpdateBehaviour(IInteractionBehaviour behaviour, Hand _hand) {
+      GrabClassifierHeuristics.GrabClassifier classifier;
+      Dictionary<IInteractionBehaviour, GrabClassifierHeuristics.GrabClassifier> classifiers = (_hand.IsLeft ? leftGrabClassifiers : rightGrabClassifiers);
+      if (!classifiers.TryGetValue(behaviour, out classifier)) {
+        classifier = new GrabClassifierHeuristics.GrabClassifier(behaviour.gameObject);
+        classifiers.Add(behaviour, classifier);
+      }
+
+      //Do the actual grab classification logic
+      fillClassifier(_hand, ref classifier);
+      GrabClassifierHeuristics.UpdateClassifier(classifier, collidingCandidates, grabParams);
+
+      if (classifier.isGrabbing != classifier.prevGrabbing) {
+        if (classifier.isGrabbing) {
+          if (!_manager.TwoHandedGrasping) { _manager.ReleaseObject(behaviour); }
+          _manager.GraspWithHand(_hand, behaviour);
+        } else if (!classifier.isGrabbing || behaviour.IsBeingGraspedByHand(_hand.Id)) {
+          _manager.ReleaseHand(_hand.Id);
+          classifier.coolDownProgress = 0f;
+        }
+      }
+      classifier.prevGrabbing = classifier.isGrabbing;
+    }
+
+    //Classifier bookkeeping
+    public void UpdateHeuristicClassifier(Hand hand) {
+      if (hand != null) {
+
+        //First check if already holding an object and only process that one
+        bool alreadyGrasping = false;
+        var graspedBehaviours = _manager.GraspedObjects;
+        for (int i = 0; i < graspedBehaviours.Count; i++) {
+          if (graspedBehaviours[i].IsBeingGraspedByHand(hand.Id)) {
+            UpdateBehaviour(graspedBehaviours[i], hand);
+            alreadyGrasping = true;
+            break;
```

Comment:
You could just return here instead of break, and do away with the bool check entirely.